design/general-progress-monitoring.md

```diff
@@ -0,0 +1,497 @@
+VolumeSnapshotter/BackupItemAction operations and call the _Progress_ method on each of them.
+
+If all snapshot and backup item operations have finished (either successfully or failed), the backup
+will be completed and the backup will move to the appropriate terminal phase and upload the
+`velero-backup.json` object to the object store and the backup will be complete.
+
+If any of the snapshots or backup items are still being processed, the phase of the backup will be
+set to the appropriate phase (_WaitingForPluginOperations_ or
+_WaitingForPluginOperationsPartialFailure_).  In the event of any of the progress checks return an
+error, the phase will move to _WaitingForPluginOperationsPartialFailure_.  The backup will then be
+requeued and will be rechecked again after some time has passed.
+
+## Restore workflow changes
+
+The restore workflow remains the same until velero would currently move the backup into one of the
+terminal states.  At this point, we will queue the restore to a finalization go-routine.  The next
+restore may then begin.  The finalization routine will run across all of the RestoreItemAction
+operations and call the _Progress_ method on each of them.
+
+If all restore item operations have finished (either successfully or failed), the restore will be
+completed and the restore will move to the appropriate terminal phase and the restore will be
+complete.
+
+If any of the restore items are still being processed, the phase of the restore will be set to the
+appropriate phase (_WaitingForPluginOperations_ or _WaitingForPluginOperationsPartialFailure_).  In
+the event of any of the progress checks return an error, the phase will move to
+_WaitingForPluginOperationsPartialFailure_.  The restore will then be requeued and will be rechecked
+again after some time has passed.
+
+## Restart workflow
+
+On restart, the Velero server will scan all Backup/Restore resources.  Any Backup/Restore resources
+which are in the _InProgress_ phase will be moved to the _Failed_ phase.  Any Backup/Restore
+resources in the _WaitingForPluginOperations_ or _WaitingForPluginOperationsPartialFailure_ phase
+will be treated as if they have been requeued and progress checked and the backup/restore will be
+requeued or moved to a terminal phase as appropriate.
+
+## Notes on already-merged code which may need updating
+
+Since this design is modifying a previously-approved design, there is some preparation work based on
+the earlier upload progress monitoring design that may need modification as a result of these
+updates. Here is a list of some of these items:
+
+1. Consts for the "Uploading" and "UploadingPartialFailure" phases have already been defined. These
+   will need to be removed when the "WaitingForPluginOperations" and
+   "WaitingForPluginOperationsPartialFailure" phases are defined.
+   - https://github.com/vmware-tanzu/velero/pull/3805
+1. Remove the ItemSnapshotter plugin APIs (and related code) since the revised design will reuse
```

Comment:
Agree, but need to check with Dave before removing it, and UploadProgressFeatureFlag removing also needs confirm.